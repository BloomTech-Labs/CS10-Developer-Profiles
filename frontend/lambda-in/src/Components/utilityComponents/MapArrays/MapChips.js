import React from 'react';
import PropTypes from 'prop-types';
import Grid from '@material-ui/core/Grid';
import DeleteForever from '@material-ui/icons/DeleteForever';
import Chip from '@material-ui/core/Chip';

/**
 * Render a list of @material-ui's Chip components.
 *
 * @function
 * @callback removeItem
 * @prop {String[]|Number[]} props - Array of Strings.
 * @prop {object} props.classes - Object defining the classes needed.
 * @prop {object} props.classes.container - Class for the 'container' Node.
 * @prop {object} props.classes.item - Class for each 'item' Node.
 * @prop {removeItem} props.removeItem - Handler to delete and item.
 * @prop {string} props.field - The user property to handle.
 * @return {object} A React component
 */

const MapChips = (props) => {
  // prettier-ignore
  const {
    field,
    classes,
    array,
    removeItem,
  } = props;

  return (
    <Grid
      container
      spacing={16}
      className={classes.container}
      alignItems="center"
      direction="row"
      justify="center"
    >
      {array.map((value, index) => (
        <div key={`${value}${Math.random()}`}>
          <Chip
            id={`${index}$${value}`}
            label={`${value}`}
            className={classes.item}
            deleteIcon={<DeleteForever />}
            onDelete={removeItem && removeItem(field, index)}
          />
        </div>
      ))}
    </Grid>
  );
};

MapChips.propTypes = {
  array: PropTypes.arrayOf(PropTypes.any).isRequired,
  classes: PropTypes.objectOf(PropTypes.string),
  removeItem: PropTypes.func,
  field: PropTypes.string,
};
MapChips.defaultProps = {
  classes: {},
  removeItem: null,
  field: null,
};

export default MapChips;
