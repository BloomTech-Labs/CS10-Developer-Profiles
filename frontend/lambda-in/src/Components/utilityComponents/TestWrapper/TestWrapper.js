import React, { Component, Fragment } from 'react';

/**
 * A Functional stateless component wrapper
 *
 * @description Enzyme's 'shallow' method don't create an instance for
 * stateless functional componets
 * Whitout an instance is not possible to access the functional componet's
 * props.
 * Because enzyme's shallow generate an instance for class coponents, <TestWrapper/>
 * allow us to access any functional coponent passed as a children.
 * @see https://github.com/airbnb/enzyme/issues/1232#issuecomment-334510754
 */
class TestWrapper extends Component {
  constructor(props) {
    super(props);
    this.state = {};
  }

  /**
   * Catch children component error an log them in test-enviroment
   *
   * @description An Error bundary to catch potential children component's
   * Javascrip errors
   * @see https://reactjs.org/docs/error-boundaries.html
   * @param {object} error - Error ebject
   * @param {object} info - Component stack on error
   */
  componentDidCatch(error, info) {
    // eslint-disable-next-line no-console
    console.log('Error Bundary', error, info);
  }

  render() {
    // eslint-disable-next-line react/prop-types
    const { children } = this.props;
    return <Fragment>{children({ ...this.props })}</Fragment>;
  }
}

export default TestWrapper;
