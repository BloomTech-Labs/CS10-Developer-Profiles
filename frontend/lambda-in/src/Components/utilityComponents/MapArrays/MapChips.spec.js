import React from 'react';
import { configure, mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import renderer from 'react-test-renderer';
import TestWrapper from '../TestWrapper/TestWrapper';
import MapChips from './MapChips';

configure({ adapter: new Adapter() });

/**
 * @author jesuarva
 */
describe('MapChips component', () => {
  // prettier-ignore
  const removeItem = jest.fn((field, index) => e => Object.assign({}, e, { field }, { index }));

  const props = {
    fullOneItem: {
      field: 'topSkills',
      array: ['JS'],
      classes: {
        container: 'container-class',
        item: 'item-class',
      },
      removeItem,
    },
    full: {
      field: 'topSkills',
      array: ['JS', 'React', 'TDD', 'Full stack developer'],
      classes: {
        container: 'container-class',
        item: 'item-class',
      },
      removeItem,
    },
    onlyArray: {
      array: ['JS', 'React', 'TDD', 'Full stack developer'],
      classes: {
        container: 'container-class',
        item: 'item-class',
      },
    },
  };

  const mapChips = {
    full: <MapChips {...props.full} />,
    fullOneItem: <MapChips {...props.fullOneItem} />,
    onlyArray: <MapChips {...props.onlyArray} />,
  };

  describe('Renders correctly', () => {
    it('sholud renders whit delete icon', () => {
      const tree = renderer.create(mapChips.full).toJSON();
      expect(tree).toMatchSnapshot();
    });
    it('sholud renders whitout delete icon', () => {
      const tree = renderer.create(mapChips.onlyArray).toJSON();
      expect(tree).toMatchSnapshot();
    });
  });

  describe('Renders only array items and no icons', () => {
    const mounted = mount(
      <TestWrapper>{() => mapChips.onlyArray}</TestWrapper>,
    );
    const items = mounted.find('span');
    const icons = mounted.find('svg');

    it('should render all items in the array', () => {
      expect(items).toHaveLength(4);
    });

    it('should no render icons', () => {
      expect(icons).toHaveLength(0);
    });
  });

  describe('Renders only array items and icons', () => {
    const mounted = mount(<TestWrapper>{() => mapChips.full}</TestWrapper>);
    const items = mounted.find('span');
    const icons = mounted.find('svg');

    it('should render all items in the array', () => {
      expect(items).toHaveLength(4);
    });

    it('should render all icons if passed as props', () => {
      expect(icons).toHaveLength(4);
    });

    it('should call `removeItem` function clicking on an icon', () => {
      icons.forEach((node) => {
        node.simulate('click');
      });

      /**
       * removeItem creates a closure: So it is called 8 times.
       * - 4 times setting up the onClick event handlers.
       * - 4 times on 'clicking' each icon.
       */
      expect(removeItem.mock.calls.length).toBe(8);
    });
  });
});
