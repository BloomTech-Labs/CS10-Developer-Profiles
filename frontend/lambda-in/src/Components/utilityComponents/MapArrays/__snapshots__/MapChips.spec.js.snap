// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MapChips component Renders correctly sholud renders whit delete icon 1`] = `
<div
  className="MuiGrid-container-1 MuiGrid-spacing-xs-16-24 MuiGrid-align-items-xs-center-9 MuiGrid-justify-xs-center-18 container-class"
>
  <div>
    <div
      className="MuiChip-root-98 MuiChip-deletable-104 item-class"
      id="0$JS"
      onClick={undefined}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      role="button"
      tabIndex={0}
    >
      <span
        className="MuiChip-label-117"
      >
        JS
      </span>
      <svg
        aria-hidden="true"
        className="MuiSvgIcon-root-123 MuiChip-deleteIcon-118"
        color={undefined}
        focusable="false"
        onClick={[Function]}
        role="presentation"
        viewBox="0 0 24 24"
      >
        <path
          d="M0 0h24v24H0V0z"
          fill="none"
        />
        <path
          d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zm2.46-7.12l1.41-1.41L12 12.59l2.12-2.12 1.41 1.41L13.41 14l2.12 2.12-1.41 1.41L12 15.41l-2.12 2.12-1.41-1.41L10.59 14l-2.13-2.12zM15.5 4l-1-1h-5l-1 1H5v2h14V4z"
        />
        <path
          d="M0 0h24v24H0z"
          fill="none"
        />
      </svg>
    </div>
  </div>
  <div>
    <div
      className="MuiChip-root-98 MuiChip-deletable-104 item-class"
      id="1$React"
      onClick={undefined}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      role="button"
      tabIndex={0}
    >
      <span
        className="MuiChip-label-117"
      >
        React
      </span>
      <svg
        aria-hidden="true"
        className="MuiSvgIcon-root-123 MuiChip-deleteIcon-118"
        color={undefined}
        focusable="false"
        onClick={[Function]}
        role="presentation"
        viewBox="0 0 24 24"
      >
        <path
          d="M0 0h24v24H0V0z"
          fill="none"
        />
        <path
          d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zm2.46-7.12l1.41-1.41L12 12.59l2.12-2.12 1.41 1.41L13.41 14l2.12 2.12-1.41 1.41L12 15.41l-2.12 2.12-1.41-1.41L10.59 14l-2.13-2.12zM15.5 4l-1-1h-5l-1 1H5v2h14V4z"
        />
        <path
          d="M0 0h24v24H0z"
          fill="none"
        />
      </svg>
    </div>
  </div>
  <div>
    <div
      className="MuiChip-root-98 MuiChip-deletable-104 item-class"
      id="2$TDD"
      onClick={undefined}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      role="button"
      tabIndex={0}
    >
      <span
        className="MuiChip-label-117"
      >
        TDD
      </span>
      <svg
        aria-hidden="true"
        className="MuiSvgIcon-root-123 MuiChip-deleteIcon-118"
        color={undefined}
        focusable="false"
        onClick={[Function]}
        role="presentation"
        viewBox="0 0 24 24"
      >
        <path
          d="M0 0h24v24H0V0z"
          fill="none"
        />
        <path
          d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zm2.46-7.12l1.41-1.41L12 12.59l2.12-2.12 1.41 1.41L13.41 14l2.12 2.12-1.41 1.41L12 15.41l-2.12 2.12-1.41-1.41L10.59 14l-2.13-2.12zM15.5 4l-1-1h-5l-1 1H5v2h14V4z"
        />
        <path
          d="M0 0h24v24H0z"
          fill="none"
        />
      </svg>
    </div>
  </div>
  <div>
    <div
      className="MuiChip-root-98 MuiChip-deletable-104 item-class"
      id="3$Full stack developer"
      onClick={undefined}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      role="button"
      tabIndex={0}
    >
      <span
        className="MuiChip-label-117"
      >
        Full stack developer
      </span>
      <svg
        aria-hidden="true"
        className="MuiSvgIcon-root-123 MuiChip-deleteIcon-118"
        color={undefined}
        focusable="false"
        onClick={[Function]}
        role="presentation"
        viewBox="0 0 24 24"
      >
        <path
          d="M0 0h24v24H0V0z"
          fill="none"
        />
        <path
          d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zm2.46-7.12l1.41-1.41L12 12.59l2.12-2.12 1.41 1.41L13.41 14l2.12 2.12-1.41 1.41L12 15.41l-2.12 2.12-1.41-1.41L10.59 14l-2.13-2.12zM15.5 4l-1-1h-5l-1 1H5v2h14V4z"
        />
        <path
          d="M0 0h24v24H0z"
          fill="none"
        />
      </svg>
    </div>
  </div>
</div>
`;

exports[`MapChips component Renders correctly sholud renders whitout delete icon 1`] = `
<div
  className="MuiGrid-container-1 MuiGrid-spacing-xs-16-24 MuiGrid-align-items-xs-center-9 MuiGrid-justify-xs-center-18 container-class"
>
  <div>
    <div
      className="MuiChip-root-98 item-class"
      id="0$JS"
      onClick={undefined}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      role="button"
      tabIndex={-1}
    >
      <span
        className="MuiChip-label-117"
      >
        JS
      </span>
    </div>
  </div>
  <div>
    <div
      className="MuiChip-root-98 item-class"
      id="1$React"
      onClick={undefined}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      role="button"
      tabIndex={-1}
    >
      <span
        className="MuiChip-label-117"
      >
        React
      </span>
    </div>
  </div>
  <div>
    <div
      className="MuiChip-root-98 item-class"
      id="2$TDD"
      onClick={undefined}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      role="button"
      tabIndex={-1}
    >
      <span
        className="MuiChip-label-117"
      >
        TDD
      </span>
    </div>
  </div>
  <div>
    <div
      className="MuiChip-root-98 item-class"
      id="3$Full stack developer"
      onClick={undefined}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      role="button"
      tabIndex={-1}
    >
      <span
        className="MuiChip-label-117"
      >
        Full stack developer
      </span>
    </div>
  </div>
</div>
`;
